File: ios/App/Store/BillingStore.swift
import SwiftUI

/// State for the payments-kernel twin surfaces: the Settings Billing card
/// (mySubscription + the Billing Portal) and the subscribe flow. State only —
/// the workflows live in BillingComponent.
@MainActor
final class BillingStore: ObservableObject {
  @Published private(set) var subscription: SubscriptionSummary?
  /// True once the first mySubscription load settled (the Billing card stays
  /// hidden until then, mirroring the web card's render-nothing-while-loading).
  @Published private(set) var hasLoadedSubscription = false
  @Published private(set) var isLoadingSubscription = false
  @Published private(set) var isOpeningPortal = false
  @Published private(set) var isStartingCheckout = false
  @Published private(set) var billingError: String?
  @Published private(set) var checkoutError: String?

  func setSubscription(_ value: SubscriptionSummary?) {
    subscription = value
    hasLoadedSubscription = true
  }

  func setLoadingSubscription(_ value: Bool) {
    isLoadingSubscription = value
  }

  func setOpeningPortal(_ value: Bool) {
    isOpeningPortal = value
  }

  func setStartingCheckout(_ value: Bool) {
    isStartingCheckout = value
  }

  func setBillingError(_ value: String?) {
    billingError = value
  }

  func setCheckoutError(_ value: String?) {
    checkoutError = value
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/store/BillingStore.kt
package com.baseapp.android.store

import com.baseapp.android.components.billing.SubscriptionSummary
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * State for the payments-kernel twin surfaces: the Settings Billing card
 * (mySubscription + the Billing Portal) and the subscribe flow. State only —
 * the workflows live in BillingComponent. Mirrors the iOS BillingStore.
 */
class BillingStore {
    data class State(
        val subscription: SubscriptionSummary? = null,
        /**
         * True once the first mySubscription load settled (the Billing card
         * stays hidden until then, mirroring the web card's
         * render-nothing-while-loading).
         */
        val hasLoadedSubscription: Boolean = false,
        val isLoadingSubscription: Boolean = false,
        val isOpeningPortal: Boolean = false,
        val isStartingCheckout: Boolean = false,
        val billingError: String? = null,
        val checkoutError: String? = null,
    )

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state.asStateFlow()

    fun setSubscription(value: SubscriptionSummary?) {
        _state.value = _state.value.copy(subscription = value, hasLoadedSubscription = true)
    }

    fun setLoadingSubscription(value: Boolean) {
        _state.value = _state.value.copy(isLoadingSubscription = value)
    }

    fun setOpeningPortal(value: Boolean) {
        _state.value = _state.value.copy(isOpeningPortal = value)
    }

    fun setStartingCheckout(value: Boolean) {
        _state.value = _state.value.copy(isStartingCheckout = value)
    }

    fun setBillingError(value: String?) {
        _state.value = _state.value.copy(billingError = value)
    }

    fun setCheckoutError(value: String?) {
        _state.value = _state.value.copy(checkoutError = value)
    }
}
